worker/src/main/java/gov/cms/ab2d/worker/processor/coverage/CoverageDriverImpl.java

```diff
@@ -73,43 +107,68 @@ public void queueStaleCoveragePeriods() {
         }
 
         log.info("queued all stale coverage periods");
+        return outOfDateInfo;
     }
 
     /**
      * Discover any nonexistent coverage periods and add them to the list of coverage periods
      */
     @Override
-    public void discoverCoveragePeriods() {
+    public void discoverCoveragePeriods() throws CoverageDriverException, InterruptedException {
 
-        log.info("discovering all coverage periods that should exist");
+        Lock lock = coverageLockWrapper.getCoverageLock();
+        boolean locked = false;
 
-        List<Contract> attestedContracts = contractService.getAllAttestedContracts();
+        try {
 
-        ZonedDateTime now = ZonedDateTime.now();
+            locked = lock.tryLock(SIXTY_SECONDS, TimeUnit.SECONDS);
 
-        for (Contract contract : attestedContracts) {
-            ZonedDateTime attestationTime = contract.getESTAttestationTime();
+            if (locked) {
+                log.info("discovering all coverage periods that should exist");
 
-            // Force first coverage period to be after
-            // January 1st 2020 which is the first moment we report data for
-            if (attestationTime.isBefore(AB2D_EPOCH)) {
-                log.debug("contract attested before ab2d epoch setting to epoch");
-                attestationTime = AB2D_EPOCH;
-            }
+                // Iterate through all attested contracts and look for new
+                // coverage periods for each contract
+                List<Contract> attestedContracts = contractService.getAllAttestedContracts();
+                for (Contract contract : attestedContracts) {
+                    discoverCoveragePeriods(contract);
+                }
 
-            int coveragePeriodsForContracts = 0;
-            while (attestationTime.isBefore(now)) {
-                coverageService.getCreateIfAbsentCoveragePeriod(contract, attestationTime.getMonthValue(), attestationTime.getYear());
-                coveragePeriodsForContracts += 1;
+                log.info("discovered all coverage periods now exiting");
+            } else {
+                throw new CoverageDriverException("could not retrieve lock to discover new coverage periods");
+            }
 
-                attestationTime = attestationTime.plusMonths(1);
+        } catch (InterruptedException interruptedException) {
+            log.error("locking interrupted so stale coverage periods could not be updated");
+            throw interruptedException;
```

Comment:
Wrap in a Runtime exception.

Reply:
That is not what I want though because we've interrupted the thread and someone needs to handle it. If I'm not handling it the interrupt flag still needs to be set or unset.

Reply:
But I'm not handling this. When this exception is called someone interrupted the thread and I need to let the parent method (with the controlling thread) know that this was interrupted. If I don't throw this then there is no way for that thread to know that it needs to clean up and shut down.